### ISSUE TITLE
Attach user to view

### ISSUE THREAD

**Issue by cromartie1984** [cyrildewit/eloquent-viewable#111]
how can i add user (if user is logged in) to views table ?

**Comment by cyrildewit** [comment]
Hey @cromartie1984,

This package doesn't relate views with users. It could be added as an feature.

**Comment by ludo237** [comment]
That can be a really nice addition

**Comment by cyrildewit** [comment]
I'm thinking about a more general solution that covers this use case, but also some others.

Something like: https://docs.spatie.be/laravel-activitylog/v3/basic-usage/logging-activity#setting-custom-properties

```php
views(Post::model)
    ->withProperties([
        'user_id' => $user->getKey(), // $user->id
    ])
    ->record();
```

But I need to dive into it to see drawbacks of this solution. One potential drawback where I can think about is for example, that it's not possible to these property data in our SQL queries.

**Comment by ludo237** [comment]
I think that the missing relationship will be one, if not the only, main drawback. 

Options are:

- `viewer` Morph Table
- JSON property

Morph table will allow a viewable to be "viewed" by any entity and it's not a constraint on the viewable itself, also it won't break any current flow AFAIK.

**Comment by sebastiaanluca** [comment]
Would you be open to a PR for this? I'd implement the (nullable) morph type variant so any model (or none) can "view" another model.

We're searching for a package that does all yours does (thanks btw 😄), but 2 caveats:
- We want to track events others than `viewed` (but that might be solvable using custom collections)
- We want to (optionally) track who performed that action

**Comment by cyrildewit** [comment]
Hi @sebastiaanluca,

Yes, I'm open to a PR for this feature, but be aware that this would need to go to the `release/v5.0.0` branch (#172).

There is already a `Viewer` class which acts as a repository for viewer related data (cookie id, ip address, check for DNT header, and crawler check). I wonder how we could combine these two concepts.

If you need some help or want to ask me something feel free to get in touch with me. 

_Could you create a new issue and explain a little bit more about the first caveat?_

Thanks,
Cyril

**Comment by kfirba** [comment]
@cyrildewit @sebastiaanluca I think that introducing more tables and "viewer" types will make the package over complicated. In my opinion, the package should stay relatively lean while offering the ability to extend it manually.

The way I solve this issue is relatively simple. I've added a `user_id` column to the migration and then created  a new `View` model that extends the original `View` model with a `creating()` hook:

```php
<?php

namespace App;

use Illuminate\Support\Facades\Auth;
use CyrildeWit\EloquentViewable\View as BaseView;

class View extends BaseView
{
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($view) {
            $view->user_id = Auth::id();
        });
    }
}
```

Then, I've replaced the binding of the `ViewContract`

```php
use App\View;
use CyrildeWit\EloquentViewable\Contracts\View as ViewContract;

public function register()
{
    $this->app->bind(ViewContract::class, View::class);
}
```

Now for the lifetime of my application, I won't need to worry about this. Let alone it is now much easier and simpler to query the database for that type of information (no need to check for polymorphic relationship).

By the way, this package is great!  :smile: 

P.S. @cyrildewit In the readme file when you show an example of queueing the process of recording a view you explicitly set a 30 minutes delay on the job. Is there a reason you did that?

Also, I think it would be more beneficial to add a built-in way to queue the jobs that will queue the job only after performing a quick check and decide whether the visit should be recorded. As it stands, for each page view we will queue a job to record the visit. Since the job is running on the queue, it does not have access to the session, hence causing the `delayInSession` to not work (unless I missed something). What do you think?

**Comment by cyrildewit** [comment]
@kfirba I never reacted to your comment, but I actually read it quite a few times. My apologies.

I agree with you on:
> I think that introducing more tables and "viewer" types will make the package over complicated.

I don't have any plans anymore for this feature.  I do still see some value in my `withProperties()` proposal. It's a more generic approach.

Another way to solve this is by extending the main `Views` class and overriding the `record()` method. There may be some improvements that can be made to make this function more extensible without copying it completely.

> In the readme file when you show an example of queueing the process of recording a view you explicitly set a 30 minutes delay on the job. Is there a reason you did that?

No, it's just a random number.

>  Since the job is running on the queue, it does not have access to the session, hence causing the delayInSession to not work (unless I missed something). What do you think?

I have added this to my to-do list for `v5.0.0`.

I will need to experiment with it more to really decide if I want it in the core. I think it's already possible to bypass that by constructing your own instance of the `ViewSessionHistory`, but that doesn't look very clean.
